api/v1beta2/provider_types.go

```diff
@@ -0,0 +1,167 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta2
+
+import (
+	"time"
+
+	"github.com/fluxcd/pkg/apis/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ProviderKind           string = "Provider"
+	GenericProvider        string = "generic"
+	GenericHMACProvider    string = "generic-hmac"
+	SlackProvider          string = "slack"
+	GrafanaProvider        string = "grafana"
+	DiscordProvider        string = "discord"
+	MSTeamsProvider        string = "msteams"
+	RocketProvider         string = "rocket"
+	GitHubDispatchProvider string = "githubdispatch"
+	GitHubProvider         string = "github"
+	GitLabProvider         string = "gitlab"
+	BitbucketProvider      string = "bitbucket"
+	AzureDevOpsProvider    string = "azuredevops"
+	GoogleChatProvider     string = "googlechat"
+	WebexProvider          string = "webex"
+	SentryProvider         string = "sentry"
+	AzureEventHubProvider  string = "azureeventhub"
+	TelegramProvider       string = "telegram"
+	LarkProvider           string = "lark"
+	Matrix                 string = "matrix"
+	OpsgenieProvider       string = "opsgenie"
+	AlertManagerProvider   string = "alertmanager"
+)
+
+// ProviderSpec defines the desired state of the Provider.
+type ProviderSpec struct {
+	// Type specifies which Provider implementation to use.
+	// +kubebuilder:validation:Enum=slack;discord;msteams;rocket;generic;generic-hmac;github;gitlab;bitbucket;azuredevops;googlechat;webex;sentry;azureeventhub;telegram;lark;matrix;opsgenie;alertmanager;grafana;githubdispatch;
+	// +required
+	Type string `json:"type"`
+
+	// Channel specifies the destination channel where events should be posted.
+	// +kubebuilder:validation:MaxLength:=2048
+	// +optional
+	Channel string `json:"channel,omitempty"`
+
+	// Username specifies the name under which events are posted.
+	// +kubebuilder:validation:MaxLength:=2048
+	// +optional
+	Username string `json:"username,omitempty"`
+
+	// Address specifies the HTTP/S incoming webhook address of this Provider.
+	// +kubebuilder:validation:Pattern="^(http|https)://"
+	// +kubebuilder:validation:MaxLength:=2048
+	// +kubebuilder:validation:Optional
+	// +optional
+	Address string `json:"address,omitempty"`
+
+	// Timeout for sending alerts to the Provider.
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ms|s|m))+$"
+	// +optional
+	Timeout *metav1.Duration `json:"timeout,omitempty"`
+
+	// Proxy the HTTP/S address of the proxy server.
+	// +kubebuilder:validation:Pattern="^(http|https)://"
+	// +kubebuilder:validation:MaxLength:=2048
+	// +kubebuilder:validation:Optional
+	// +optional
+	Proxy string `json:"proxy,omitempty"`
+
+	// SecretRef specifies the Secret containing the authentication
+	// credentials for this Provider.
+	// +optional
+	SecretRef *meta.LocalObjectReference `json:"secretRef,omitempty"`
```

Comment:
If the secret gets deleted later on, the controller will not notice it, since we don't have watches for Secrets/ConfigMaps. A miss configured secret will be logged, when an alert or receiver fails, so you would diagnose these errors the same way as with wrong credentials, the controller does not check them, they get validated once an event is handled.

Reply:
> the controller does not check them, they get validated once an event is handled

The Provider reconciler has this notification factory code https://github.com/fluxcd/notification-controller/blob/a8f8dfccffedccd8cc7ebc128b4369e7c5fd2e9c/controllers/provider_controller.go#L209-L212 which is similar to the code in the event handler https://github.com/fluxcd/notification-controller/blob/a8f8dfccffedccd8cc7ebc128b4369e7c5fd2e9c/internal/server/event_handlers.go#L246-L247 which actually checks if the provided configuration is valid.

Reply:
If we issues events on every error we’ll end up DDOSing the Kubernetes API with events, what if you receive thousands of calls to the receiver endpoints from some buggy integration?

Reply:
Is that any different from what's happening today? When a Receiver receives a payload, it annotates the respective Resource object, which may be more expensive compared to just emitting an event. Thousands of calls to receiver endpoint will also result in k8s API calls for annotating the resource, if I'm not misunderstanding the code.

For the event handler, I know there's some kind of de-duplication, can't find the code now, which prevents duplicate events from resulting in duplicate notification. This de-duplication may help prevent too many k8s events for the same event. This is also how I believe we prevent spamming the provider endpoint.

Similarly, if we add some mechanism to prevent receiver spams, that may also help with avoiding event spams.

Reply:
There is nothing to annotate if the handshake fails. Also writing to etcd thousands of new objects is very different from  issuing thousands of patches for the same object.

Currently we don't have a rate limiter for receivers, as stated in docs, we expect people to setup rate limiting at ingress level. IMO we are better off adding an interval (with a 60s default) to receivers and rerun the validation instead of writing events from the web server.

Reply:
> There is nothing to annotate if the handshake fails.

You mean the handshake between the external request and the receiver server, right? If that is so, I think that's failing too early. We may not have the payload to determine which Receiver the request belongs to. So we can't annotate or record event in such situation.

If we expect users to have their own rate limiting solution, then we don't have the event spam issue anymore.

> we are better off adding an interval (with a 60s default) to receivers

Just the Receiver? Not the Provider and Alert? It seems like none of these objects are static and have dynamic behavior because of their dependency on other resources. If they reflect their actual current state, that'll help prevent requests in the web server which are known in advance to fail because the Receiver, Provider or Alert aren't ready.

Reply:
Also, maybe Alert should watch Provider, similar to how ImagePolicy watches ImageRepository in image-reflector-controller, as it's strictly dependent on Provider. No need of interval in Alert.

Reply:
> If we issues events on every error we’ll end up DDOSing the Kubernetes API with events, what if you receive thousands of calls to the receiver endpoints from some buggy integration?

We probably should have a back-off/kill-switch mechanism to reject calls that we know won't be able to be processed, and to decrease the impact on downstream systems.

Although we should do our part and make sure that NC handles this properly, it is worth bearing in mind that as a last line of defence production environments are meant to have `EventRateLimit` enabled and configured, which would decrease the likelihood of apiservers being DOS'd.

Reply:
I've added `spec.interval` to periodically reconcile providers and receivers with their Secret references to surface config errors after initialisation. Alerts already watch for Provider changes.

Reply:
> Also writing to etcd thousands of new objects is very different from issuing thousands of patches for the same object.

This is simply not what is happening within the Kubernetes event broadcaster.

- https://github.com/kubernetes/client-go/blob/master/tools/record/event.go#L207
- https://github.com/kubernetes/client-go/blob/master/tools/record/events_cache.go#L238-L241
- https://github.com/kubernetes/client-go/blob/master/tools/record/events_cache.go#L429-L438

From the latter link:

> A source may burst 25 events about an object, but has a refill rate budget per object of 1 event every 5 minutes to control long-tail of spam.

Reply:
OK so this is even worse for the controller, the events will pile up in memory and the go routines will be stuck in sleep.

Reply:
I propose we tackle the events issues by web servers in a different PR/issue. The scope of this PR is to refactor the reconcilers, let's create an issue for web servers and error handling, if and how we duplicate logs into events for those.